compiler/lib/effects.ml

```diff
@@ -0,0 +1,552 @@
+      binding the corresponding one-level stack to [stack] and the
+      low-level continuation of the next fiber to [k']. *)
+
+  val resume_stack : stack -> cont -> instr list * cont
+  (** [resume_stack stack k] returns [(instr, k')], where [instrs] is
+      the list of instructions necessary to updates the execution
+      context with the stack of fibers in [stack] (so as to resume a
+      continuation), binding the low-level continuation of the
+      innermost fiber to [k'] *)
+
+  val cons_fiber : stack -> continuation -> instr list
+  (** [cons_fiber stack cont] returns the list of instructions
+      necessary to append a one-level stack [stack] to the stack of
+      fibers of the continuation. *)
+
+  val get_effect_handler : stack -> instr list * Var.t
+  (** [get_effect_handler stack] returns [(instrs, handler)], where
+      [instrs] is the list of instructions necessary to bind the
+      topmost effect handler in [stack] to [handler]. *)
+
+  val continuation_of_stack : stack -> instr list * continuation
+  (** [continuation_of_stack stack] returns [(instrs, handler)], where
+      [instrs] is the list of instructions necessary to allocate a
+      continuation corresponding to stack [stack] and bind it to
+      [continuation] *)
+
+  val push_trap : cont -> instr list
+  (** [push_trap exn_handler] returns the list of instructions
+      necessary to push the exception handler [exn_handler] to the
+      stack of exception handlers. *)
+
+  val pop_trap : unit -> instr list * cont
+  (** [pop_trap exn_handler] returns [(instrs, exn_handler)], where
+      [instrs] is the list of instructions necessary to remove and
+      return the topmost element of the stack of exception handlers,
+      and bind it to [exn_handler]. *)
+end = struct
+  type cont = Var.t
+
+  type continuation = Var.t
+
+  type stack = Var.t
+
+  let pop_fiber k =
+    let res = Var.fresh () in
+    let stack = Var.fresh_n "stack" in
+    let k' = Var.fresh_n "cont" in
+    ( [ Let (res, Prim (Extern "caml_pop_fiber", [ Pv k ]))
+      ; Let (stack, Field (res, 0))
+      ; Let (k', Field (res, 1))
+      ]
+    , stack
+    , k' )
+
+  let resume_stack stack k =
+    let k' = Var.fresh_n "cont" in
+    [ Let (k', Prim (Extern "caml_resume_stack", [ Pv stack; Pv k ])) ], k'
+
+  let cons_fiber fiber continuation =
+    (* Assumes a one-level stack (the last component of [fiber] is empty). *)
+    (* let Cons (k, exn_stack, handlers, Empty) = fiber in
+       continuation := Cons (k, exn_stack, handlers, !continuation) *)
+    let stack = Var.fresh () in
+    let k = Var.fresh () and exn_stack = Var.fresh () and b = Var.fresh () in
+    let handlers = Var.fresh () in
+    [ Let (stack, Field (continuation, 0))
+    ; Let (k, Field (fiber, 0))
+    ; Let (exn_stack, Field (fiber, 1))
+    ; Let (handlers, Field (fiber, 2))
+    ; Let (b, Block (0, [| k; exn_stack; handlers; stack |], NotArray))
+    ; Set_field (continuation, 0, b)
+    ]
```

Comment:
Why do we inline instructions instead of solely relying on primitive calls ?

Reply:
There are only three occurrences of`%reperform`, so the code size was not critical.

For `%perform`, you need to call the effect handler from the primitive if you want a significant size gain. Otherwise, the primitive returns many results that you have to extract using inline instructions.